### DISCUSSION TITLE
Does libcurl support specifying a different name from URL for TLS SNI?

### DISCUSSION THREAD

**Discussion by freebendy** [curl/curl#15804] [General]
Since my application has its own DNS resolve, the host is resolved before and a ip address is set with CURLOPT_URL. In this case, libcurl won't have SNI in TLS ClienHello message and the server reset the connection. Is there a way I can specify the name for SNI?

I tried the following way, but the sni_callback() is never called.
```
CURLcode ssl_ctx_callback(CURL *curl, void *sslctx, void *parm)
{
    SSL_CTX* ctx = static_cast<SSL_CTX*>(sslctx);
    // Register the SNI callback function
    SSL_CTX_set_tlsext_servername_callback(ctx, sni_callback);
    SSL_CTX_set_tlsext_servername_arg(ctx, parm);
    return CURLE_OK;
}

int sni_callback(SSL *ssl, int *ad, void *arg)
{
    // Set the SNI for the SSL connection
    if (SSL_set_tlsext_host_name(ssl, static_cast<const char *>(arg)) != 1)
    {
        return SSL_TLSEXT_ERR_ALERT_FATAL;
    }
    return SSL_TLSEXT_ERR_OK;
}
.... 
const char sni_hostname[] = "example.com";

curl_easy_setopt(curl, CURLOPT_SSL_CTX_FUNCTION, ssl_ctx_callback);
curl_easy_setopt(curl, CURLOPT_SSL_CTX_DATA, (void *)sni_hostname);
```

Is what I want to achieve feasible in libcurl?

**Comment by bagder** [comment]
A common approach is to use the correct name in the URL then provide its IP address(es) with CURLOPT_RESOLVE.

**Comment by dfandrich** [comment]
The traditional answer for this is to use CURLOPT_RESOLVE to use a specific IP
address instead of rewriting the URL. SNI will then work correctly
automatically.

**Comment by freebendy** [comment]
Thanks for all your replies, it works!!
